### ISSUE TITLE
Opinion Needed: More buttons

### ISSUE THREAD

**Issue by ArturKovacs** [ArturKovacs/emulsion#80]
Currently there are many things that are somewhat held back or compromised by the fact that Emulsion has only two buttons on its UI. I do believe that minimalism is a feature but I think that adding just two more buttons to the bottom panel would retain the minimalistic quality while opening up many new possibilities. My current plan for those two buttons are: "Display Folder Path" on the left and "More Options" on the right.

![More Emulsion Buttons](https://user-images.githubusercontent.com/8320264/81493909-c5db5900-92a4-11ea-924c-efb949090800.png)

When clicling on "More Options" the following little panel would appear. On this panel, the buttons from left to right are the following: "Current Animation Playback Framerate", "Copy to Clipboard", "Open with External", "Browse External Program for Open with External"

![More Emulsion Buttons-Extra](https://user-images.githubusercontent.com/8320264/81493914-d12e8480-92a4-11ea-9e21-88279ebf92e7.png)

Because my opinion may not reflect what's right, I'd like to call for feedback on this. A simple thumbs up or thumbs down is already very helpful but if you don't agree with this approach please describe the alternative that you think would be better.

(This is related to #8 and #79)

**Comment by CorneliusCornbread** [comment]
Personally I don't see an issue with this, but some users might find it annoying to have to hit a pop-up before they can do any of these actions, or might find it annoying that some basic features are hidden behind a pop-up. Just some food for thought

**Comment by ArturKovacs** [comment]
I absolutely agree. I'm not sure if all of these should have a keyboard shortcut, but copying to clipboard should definitely have one, and probably the "open with external" as well. The difficulty is that there is no-one-size-fits-all layout becuse there are different workflows and use cases which might require different functionalties to be easily accessible. So my reasoning for putting the show folder path button to the bottom panel by default is that when clicked it does something that's immediately visible and easy to understand. It will make a panel appear at the top, similar to the bottom-panel, showing the full path to the current folder. All of the others either don't seem to do anything when clicked of juyst confusing for someone who doesn't know what they do. So in order to avoid making uses feel that "this program is confusing", the more advanced features are behind a submenu.

I also think there could be an option for customizing which buttons appear at the bottom. But that is out of this issue's scope.

**Comment by Shnatsel** [comment]
**TL;DR:** drop all copying/opening options other than "Open with..." (assuming that also allows opening in file browser, which on my Linux it does, not sure about Windows/Mac), drop animation controls, add GUI elements for previous/next image, original size and fit to screen.

"Browse external program for open with external" sounds complex and very niche. I'm not sure having a button for that is worthwhile. I'd just fire up the terminal or file manager if I needed that. "Display folder path" also sounds like a duplicate of that - just point me to the file in file manager and I'll take it from here.

Animation playback rate also sounds like a very niche feature. Sure, it's a cool tech demo, but if I want to quickly scroll through a large folder I already have the control element for that, and if I want to make a sequence of images into an animation I'd turn it into an APNG or a video and view that through the normal means. So I'd advocate for animation playback to be CLI-only.

On the other hand, next/previous buttons are conspicuous by their absence. It's even weirder not to have buttons for "fit to window" and "original size". Keyboard shortcuts for that are nifty, but it's not something you can expect non-technical people to memorize, especially when they're unguessable (e.g. original size being on Q and not on 1).

When it comes to UI/UX design, I cannot think of a better example to follow than the [elementary OS human interface guidelines](https://elementary.io/docs/human-interface-guidelines#human-interface-guidelines). This is a document that clearly and concisely lays out many of the principles that define good UX design.

For reference, this is what their viewer/editor looks like (although it doesn't incorporate all the principles):

![Снимок экрана от 2020-05-27 01-31-18](https://user-images.githubusercontent.com/291257/82960263-3e872880-9fba-11ea-94ec-16625a1c6452.png)

([image from wikipedia](https://commons.wikimedia.org/wiki/Commons:Picture_of_the_day#/media/File:Sun_over_Lake_Hawea,_New_Zealand.jpg))

Previous/Next, "Original size" and "Fit to window" are all present as GUI controls; there is also a zoom level widget, and that's literally it - every other control on the screen is for editing. The zoom level widget is admittedly awkward - I'd much prefer mouse wheel zoom or two-finger pan on touchscreen, without the extra UI clutter. Notably, "Open with..." is in the context menu (the popup you get by right-clicking the image) and it also contains the "Open in file browser" action, simply because the OS ships it as an "Open with" option for everything. There is no menu bar or hamburger menu.

**Comment by ArturKovacs** [comment]
>
> "Browse external program for open with external" sounds complex and very niche. [...]

I think we agree that Emulsion is not for the average user. The average user doesn't want a minimal image viewer, they probably want to be able to adjust the colours, crop, and rotate. So Emulsion is by design not for that audience. I think that Emulsion is for two kinds of people: those who don't care that much about editing the image and those who are comfortable using a full-featured image editor so they don't need any of that in an image viewer.

Considering this target audience, allowing to open the target image with a custom program (like an editor) is definitely not niche neither is the option to choose that program from within Emulsion.

Let me clarify how I planned this to work. When you click on "Open With" for the first time you get presented with a os-specific file browser dialogue (this should be reasonably easy to implement). In this dialogue window you can choose a program to open the image with. Whenever you click on "Open With" in the future the previously chosen program will be used without requesting user input. Clicking on "Browse External Program for Open with External" would bring up that dialogue again so you can change the program.

> I'd just fire up the terminal or file manager if I needed that. "Display folder path" also sounds like a duplicate of that

See #8 for a use case where "Display folder path" is not a duplicate of showing in the file browser.

> Animation playback rate also sounds like a very niche feature. [...]

When rendering an animation with software like Blender you have to choose if you want to render into separate images or into a video file. The problem with choosing a video is that if anything goes wrong in the 12 hour render time you'll probably have to start over. Due to this, it's a good idea to render into a sequence of separate image files instead, but that's not as easy to preview as a video file except if you have Emulsion. So I think this is an important feature even though it's completely useless for a portion of the userbase.

> On the other hand, next/previous buttons are conspicuous by their absence. It's even weirder not to have buttons for "fit to window" and "original size". 

This is something I'm rather torn about. I sort-of agree with you, I can see how many would like to see these buttons available. But on the other hand I can't come up with a good layout where there's at most four buttons on the bar and navigation or image sizing is included. Suggestions are welcome.

> 
> drop all copying/opening options other than "Open with..." 

Copy would copy the image contents to the clipboard. This functionality is very different from "open with" so I think both should be possible to do. For example I didn't think about setting "open with" to the file browser, I was thinking about setting it to an image editor but both are completely valid. So let's say that I have "open with" set to an image editor. When I'm browsing a folder with a bunch of images and I found one that I'd like to paste into a discord conversation, I could use "copy" to copy the image contents and then paste it into discord. So I think these both should have a GUI button.

> When it comes to UI/UX design, I cannot think of a better example to follow than the elementary OS human interface guidelines. 

I haven't finished reading it but it is good indeed.

**Comment by Shnatsel** [comment]
Something like this, perhaps?

![emulsion-mockup-4](https://user-images.githubusercontent.com/291257/83061966-1f41d700-a05e-11ea-9296-27ea08c71d3d.png)

I'm not convinced that remembering the last used program for "Open with..." is a good idea. It only takes less clicks (1 instead of 2) if the button is always present, but then you have to place the button for changing the program in a visible location too, otherwise it's hard to discover and changing the program would take a ridiculous amount of clicks. Plus there's no indication which program will be launched, so you might click it only to find it was the wrong one and have to close it and go back to change it. Thus "Open with" that always shows a dialog sounds like a good middle ground - one button, 2 clicks for any program, consistently. Incidentally that's what absolutely everyone else seems to be doing.

**Comment by ArturKovacs** [comment]
You are right. Optimally Open with should show a "drop down" menu or short list with previously selected programs and an option to browse for a program. There is one slight issue with this. Emulsion can't yet do font rendering which makes this feature impossible to implement at the current state. Showing a file browse dialogue every time you click on open with external doesn't seem something like users would like.

I do like your proposal for the button layout. For some reason I thought that adding more than four buttons would start to introduce clutter but this space between the middle part and the buttons on the sides seems it could work. I'm thinking that placing the "next" "previous" buttons right next to each other would be more comfortable to use for those who prefer the ui buttons. Thanks a lot for engaging in the conversation, this made me excited to start working on the copy to clipboard stuff so that we can push this bad boy forward. (That clipboard thing is gonna take sometime though but there's still plenty of time for the next relealse and we can re-schedule anyways.)

**Comment by Shnatsel** [comment]
I'm not sure about other platforms, but on my Linux the "Open with..." thing is clearly a reusable element - I see it all over the place. Although only the file browser has an option to use a new and weird one that's not currently listed. Example from native image viewer:

![open_with](https://user-images.githubusercontent.com/291257/83068697-038ffe00-a069-11ea-96a7-fb1b85e93c69.png)

I suspect that on Linux it lists the .desktop files that match the mime type of the image and presents that. The result is rendered in what is essentially the right-click menu. I'm not sure if there's a cross-platform way to do that. I suppose you could simply show application icons instead of text and sidestep the trouble with text rendering, but on Linux application icons are allowed to be SVG and I'm not sure if there's a raster cache of them anywhere.

I thought that copying the path is redundant to opening in file browser because in my file browser Ctrl+C doubles for copying the path. I'm not sure if that's the case on other platforms.

**Comment by ArturKovacs** [comment]
I'm going to have to look into whether the "open with" menu can be done through some system level API on other platforms.

> I thought that copying the path is redundant to opening in file browser because in my file browser Ctrl+C doubles for copying the path. 

What I meant by "copy the image contents" is that the pixel data would be placed onto the clipboard. Overall I'm not sure which one is better, it may even be possible to provide both the image data and the file url at once but implementing adding image data to the clipboard for all platforms is complicated enough so I think I'll just stick with that.
